src/Analyzers/CSharp/Tests/UseUTF8StringLiteral/UseUTF8StringLiteralTests.cs

```diff
@@ -0,0 +1,596 @@
+        public async Task TestNotInExpressionTree()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+using System;
+using System.Linq.Expressions;
+
+public class C
+{
+    public void M()
+    {
+        N(() => new byte[] { 65, 66, 67 });
+    }
+
+    public void N(Expression<Func<byte[]>> f)
+    {
+    }
+}",
+                LanguageVersion = LanguageVersion.Preview
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseUTF8StringLiteral)]
+        public async Task TestNotWhenNotByteArray()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+public class C
+{
+    public void M()
+    {
+        var x = new int[] { 65, 66, 67 };
+    }
+}",
+                LanguageVersion = LanguageVersion.Preview
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseUTF8StringLiteral)]
+        public async Task TestNotWhenOptionNotSet()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+public class C
+{
+    public void M()
+    {
+        var x = new byte[] { 65, 66, 67 };
+    }
+}",
+                EditorConfig = @"
+[*.cs]
+csharp_style_prefer_utf8_string_literal = false
+",
+                LanguageVersion = LanguageVersion.Preview
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseUTF8StringLiteral)]
+        public async Task TestNotWhenNonLiteralElement()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+public class C
+{
+    public void M()
+    {
+        var x = new byte[] { 65, GetB(), 67 };
+    }
+
+    public byte GetB() => 66;
+}",
+                LanguageVersion = LanguageVersion.Preview
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseUTF8StringLiteral)]
+        public async Task TestSimpleByteArray()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+public class C
+{
+    public void M()
+    {
+        var x = {|IDE0230:new byte[] { 65, 66, 67 }|};
+    }
+}",
+                FixedCode =
+@"
+public class C
+{
+    public void M()
+    {
+        var x = ""ABC""u8;
+    }
+}",
+                CodeActionValidationMode = CodeActionValidationMode.None,
+                LanguageVersion = LanguageVersion.Preview
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseUTF8StringLiteral)]
+        public async Task TestConstant()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode =
+@"
+public class C
+{
+    private const byte B = 66;
+    public void M()
+    {
+        var x = {|IDE0230:new byte[] { 65, B, 67 }|};
```

Comment:
Is there a summary of how the IDE will differentiate (if at all) `new byte[] { binary blob }` from `new byte[] { ascii or utf-8 string data }`?

CC. @stephentoub who may also have more insight into patterns from https://github.com/dotnet/runtime/pull/67733

Reply:
we might want some sort of threshold here.  Or just disallow Unicode.Format/Control.  I do imagine that if it's really just raw binary, then it will fail David's test to convert these to runes and so we won't offer.

Reply:
The easy cases were like:
```C#
new byte[] { (byte)'H', (byte)'e', (byte)'l', (byte)'l', (byte)'o' }
```
While it's of course possible these were actually binary blobs, I think the intent of the developer was fairly clear and it'd be quite reasonable to replace those.
e.g. https://github.com/dotnet/runtime/pull/67733/files#diff-336d569f6d3f8a493b96a64886e8900657ec8174f56c057ed34ec943e2fa1c8eL13

Other easy cases were:
```C#
Encoding.UTF8.GetBytes("constantstringofvaliddata")
```

Much less so when it's just:
```C#
new byte[] { number, number, number, ... }
```
In these cases, I don't know that an automated tool can reasonably make a distinction, e.g. this case really required some context applied:
https://github.com/dotnet/runtime/pull/67733/files#diff-5d354d9e17c353eb57b4a4b376afe1f28005f9c5140bef8017a17ff39f7e7516L8

Reply:
I'd think `tab`, and `newline` are fine (`\t`, `\r`, `\n`).

I'd think `null` (`\0`) as the final character would be fine (if C# doesn't change to implicitly put this at the end of the embedded metadata)

I'd expect `\a`, `\b`, `\f`, and `\v` to be relatively rare and potentially showcases of binary blobs. Same with other control characters.

I don't have any reasonable expectations around other Unicode sequences. Globalized scenarios could have basically any character as "common"

Reply:
> The easy cases were like:

True, feels like if you always had char literals to begin with, we should always allow that (and should just preserve whatever teh char literals contained within them).

Reply:
Thank you for starting this conversation, definitely something that needs resolving. I was thinking of only offering this if the bytes represent ASCII characters, though even that is broad enough to have oddities.

I don't see much value in replace a byte array with a bunch of `\uXXXX`s in a string, but obviously unprintable chars aren't good either.

Reply:
> e.g. this case really required some context applied:
> [dotnet/runtime#67733 (files)](https://github.com/dotnet/runtime/pull/67733/files#diff-5d354d9e17c353eb57b4a4b376afe1f28005f9c5140bef8017a17ff39f7e7516L8)

@stephentoub can you link to this directly, outside of the PR? GitHub is not taking me to anything particularly interesting when I click that link, guessing its in a file that is not expanded at first.

> Much less so when it's just: `new byte[] { number, number, number, ... }`

You mean not everyone has committed parts of the ASCII table to memory? Kids these days!

Reply:
@davidwengier The comment link is referring to `Utf8Constants.cs` line 8 :)